core/auth/src/main/java/software/amazon/awssdk/auth/credentials/ProfileCredentialsProvider.java

```diff
@@ -140,16 +143,78 @@ public String toString() {
 
     @Override
     public void close() {
-        // The delegate credentials provider may be closeable (eg. if it's an STS credentials provider). In this case, we should
-        // clean it up when this credentials provider is closed.
-        IoUtils.closeIfCloseable(credentialsProvider, null);
+        credentialsCache.close();
     }
 
     @Override
     public Builder toBuilder() {
         return new BuilderImpl(this);
     }
 
+    private AwsCredentialsProvider createCredentialsProvider(ProfileFile profileFile, String profileName) {
+        // Load the profile and credentials provider
+        return profileFile.profile(profileName)
+                          .flatMap(p -> new ProfileCredentialsUtils(profileFile, p, profileFile::profile).credentialsProvider())
+                          .orElseThrow(() -> {
+                              String errorMessage = String.format("Profile file contained no credentials for " +
+                                                                  "profile '%s': %s", profileName, profileFile);
+                              return SdkClientException.builder().message(errorMessage).build();
+                          });
+    }
+
+    private boolean needToReloadCredentials() {
+        return Objects.isNull(credentials.get()) || profileFile.get().isStale();
+    }
+
+    private RefreshResult<AwsCredentials> refreshCredentials() {
+        if (loadException != null) {
+            throw loadException;
+        }
+
+        Instant now = clock.instant();
+        Instant staleTime;
+        AwsCredentials nextCredentials;
+        if (needToReloadCredentials()) {
+            synchronized (this) {
+                ProfileFile previousProfileFile = profileFile.get();
+                ProfileFile newProfileFile = previousProfileFile.reload();
+                profileFile.set(newProfileFile);
+
+                AwsCredentialsProvider credentialsProvider = createCredentialsProvider(newProfileFile, this.profileName);
+                nextCredentials = credentialsProvider.resolveCredentials();
+                credentials.set(nextCredentials);
+
+                // The delegate credentials provider may be closeable (eg. if it's an STS credentials provider). In this case,
+                // we should clean it up when this credentials provider is closed.
+                IoUtils.closeIfCloseable(credentialsProvider, null);
+            }
+            staleTime = staleTime(now);
+        } else {
+            nextCredentials = credentials.get();
+            staleTime = pollTime(now);
+        }
+
+        return RefreshResult.builder(nextCredentials)
+                            .staleTime(staleTime)
+                            .build();
+    }
```

Comment:
There's a lot going on here that should be simplified once the reloading functionality is moved to a separate class. That way, you don't need to continuously update the `profileFile`. Instead, you can maintain a final reference to the file loader, and load the file whenever you need.

Reply:
This is similar to something we did recently with Bearer token support: https://github.com/aws/aws-sdk-java-v2/blob/58451426c75522cb26db2e48f10a1d225c93e6c8/services/ssooidc/src/main/java/software/amazon/awssdk/services/ssooidc/SsoOidcTokenProvider.java#L70

which required refreshing credentials from a file

Reply:
Thanks for the example. This helped a lot.